[agent] Add documentation for Agent encryption at rest (#2368)
elastic/observability-docs: docs/en/ingest-management/elastic-agent/elastic-agent-encryption.asciidoc

[elastic-agent-encryption]

> = Elastic Agent Configuration Encryption

Reviewer: ```suggestion
= {agent} configuration encryption
```

> It is important for users to understand the security model of Elastic Agent and how it handles sensitive values in integration configurations.

Reviewer: ```suggestion
It is important for you to understand the {agent} security model and how it handles sensitive values in integration configurations.
```

> At a high level, Elastic Agent receives configuration data from Fleet Server over an encrypted connection, persists those secrets in an encrypted form on disk.

Reviewer: ```suggestion
At a high level, {agent} receives configuration data from {fleet-server} over an encrypted connection, and persists the configuration in an encrypted form on disk.
```
Do you mean persists the configuration? "persists those secrets" makes it sound like only sensitive info is encrypted.

> This persistence is necessary to allow agents to continue to operate, even in the event of a loss of connectivity to the Fleet server.

Reviewer: ```suggestion
This persistence allows agents to continue to operate even if they are unable to connect to the {fleet-server}.
```

> The entirety of the Fleet Agent Policy is encrypted at rest, but is recoverable if the user has access to both the encrypted configuration data and the associated key.

Author: Can we double-confirm this @cmacknz ?

Reviewer 2: @aleksmaus would be good to have you review this as well.

> The entirety of the Fleet Agent Policy is encrypted at rest, but is recoverable if the user has access to both the encrypted configuration data and the associated key.

Reviewer: ```suggestion
The entire Fleet Agent Policy is encrypted at rest, but is recoverable if you have access to both the encrypted configuration data and the associated key.
```

> The key material in stored in an OS-dependent manner as described below.

Reviewer: ```suggestion
The key material is stored in an OS-dependent manner as described in the following sections.
```

[discreet]

> == Darwin (Mac OS)

Reviewer: ```suggestion
== Darwin (macOS)
```

Key material is stored in the system keychain. The value is stored as is without any additional transformations.

> [discreet]

Reviewer: ```suggestion
[discrete]
```

Reviewer: IMO asciidoc's naming here is unfortunate. As it is, you end up with:

![image](https://user-images.githubusercontent.com/14206422/202062678-8a098e05-0600-40c7-80d0-b56c07684a68.png)

== Windows

Configuration data is encrypted with https://learn.microsoft.com/en-us/dotnet/standard/security/how-to-use-data-protection[DPAPI] `CryptProtectData` with `CRYPTPROTECT_LOCAL_MACHINE``. 
Additional entropy is derived from crypto/rand bytes stored in the `.seed` file.

> Configuration data is stored as separate files where the name of the file is SHA256 hash of the key and the content of the file is encrypted with DPAPI data.

Reviewer: ```suggestion
Configuration data is stored as separate files, where the name of the file is a SHA256 hash of the key, and the content of the file is encrypted with DPAPI data.
```

The security of key data relies on file system permissions. Only the Administrator should be able to access the file.

> [discreet]

Reviewer: ```suggestion
[discrete]
```

== Linux

> The encryption key is derived from crypto/rand bytes that is written to the `.seed` file after PBKDF2 transformation.

Reviewer: ```suggestion
The encryption key is derived from crypto/rand bytes stored in the `.seed` file after PBKDF2 transformation.
```
Edited to be consistent with previous sentence about the seed file.

> Configuration data is stored as separate files where the name of the file is SHA256 hash of the key and the content of the file is AES256-GSM encrypted.

Reviewer: ```suggestion
Configuration data is stored as separate files, where the name of the file is a SHA256 hash of the key, and the content of the file is AES256-GSM encrypted.
```
